compiler/optimizations/gpuTransforms.cpp

```diff
@@ -1581,48 +1579,62 @@ void GpuKernel::generateOobCondNoIPT(Symbol* localUpperBound) {
  *
  * // recall index[0] = lowerBound + `global thread idx` * itersPerThread
  * def threadBound = min(index[0] + itersPerThread - 1, upperBound)
- *
- * CForLoop(; index[0] <= threadBound; for all i: index[i] += 1) {
+ * for(; index[0] <= threadBound; for all i: index[i] += 1) {
  *   // empty block stored in this->userBody_
  * }
+ *
+ * If we turn the for-loop into CForLoop, it will be treated as a CPU-only
+ * loop in cleanupForeachLoopsGuaranteedToRunOnCpu() that will come later
+ * and replace ex. GPU primitives like "gpu threadIdx x" with chpl_error.
+ * Not to mention that the latter will break codegen.
+ * So use WhileDoStmt instead as follows:
+ *
+ * def whileVar = index[0] <= threadBound
+ * WhileDoStmt(whileVar, {
+ *   {  // empty block stored in this->userBody_
+ *   }
+ *   for all i: index[i] += 1;
+ *   whileVar = index[0] <= threadBound
+ * })
```

Comment:
I find it surprising that the AST you add to a _kernel_ as per this comment is getting fed through `cleanupForeachLoopsGuaranteedToRunOnCpu` -- that would require the C for loop to be outside the body of the kernel, no? So rather than using a `while` loop, I think I'd prefer to ensure `cleanupForeachLoopsGuaranteedToRunOnCpu` is not called inappropriately on what appears to be kernel-only AST.